### PR TITLE
Refactor GET /scans

### PR DESCRIPTION
**Pull request by swashcap**
* **Problem:** The `GET /scans` route is a little confusing: all logic lies within the route file, instead of the controller.
* **Solution:** Move some functionality to the controller to clean it up the route file
* **Testing:**
  * Look over the commits and changed files to ensure functionality remains intact
  * Do the CI things: run `npm run lint` and `npm test` to ensure this doesn’t break the server

### CODE REVIEW

**Comment by drewlandis** [comment]
Please fix:
- [ ] merge conflicts
- [ ] rebase branch so that compiled files/linting changes don't show up in the diff
- [ ] npm test is throwing three errors on `scan` route tests

**Comment by swashcap** [comment]
@drewlandis Apologies for the confusing history! This should now be addressed. I also updated the PR description with some more details. Let me know what you think!

File: packages/steelpenny/src/lib/routes/scans.js [hunk 1]
```diff
@@ -60,28 +37,39 @@ exports.register = (server, options, next) => {
...
+          .readAll(credentials);
+
+        const bluebird = require('bluebird'); // eslint-disable-line
+
+        /**
+         * Get the studies that the user is allowed to read scans from
+         * @return {Promise} a promise that resolves to an array of studyIds
+         */
+        const getReadScanStudies = username => bluebird.promisify(
+          relations.study)('What can %s read_Scan from?', username);
```
Comment (drewlandis) [inline packages/steelpenny/src/lib/routes/scans.js]: Prefer this:
```
const getReadScanStudies = username =>
  bluebird.promisify(relations.study)('What can %s read_Scan from?', username);
```

File: packages/steelpenny/src/lib/routes/scans.js [hunk 2]
```diff
@@ -60,28 +37,39 @@ exports.register = (server, options, next) => {
...
+         * get all scans in the specified studies
+         * @param  {number} studies array of studyIds (int)
+         * @param  {boolean} returnJson return JSON instead of raw collection
+         * @return {Promise} A Promise that resolves to json or collection of scans
+         */
+        const readAllScansInStudies = (credentials, studies) => new Scan()
+          .query('whereIn', 'study_id', studies)
+          .readAll(credentials);
+
+        const bluebird = require('bluebird'); // eslint-disable-line
```
Comment (drewlandis) [inline packages/steelpenny/src/lib/routes/scans.js]: Do not eslint disable the entire line.  Please add a reason for disabling.

File: packages/steelpenny/src/lib/controllers/scans.js [hunk 3]
```diff
@@ -47,3 +48,27 @@ module.exports.scansSchema = joi.object().keys({
   projectNumber: joi.any().required(),
   chargeCode: joi.any().required(),
 }).label('Scan');
+
+/**
```
Comment (rssk) [inline packages/steelpenny/src/lib/controllers/scans.js]: This docbloc is kinda useless, be nice to have better info on this function

File: packages/steelpenny/src/lib/routes/scans.js [hunk 4]
```diff
@@ -1,8 +1,9 @@
 'use strict';
 
+const _ = require('lodash');
 const boom = require('boom');
 const joi = require('joi');
-const _ = require('lodash');
+const isEmpty = require('lodash/isEmpty');
```
Comment (swashcap) [inline packages/steelpenny/src/lib/routes/scans.js]: Updated in 3acbed1241f7a100484113aa7a28cfee24082710.

File: packages/steelpenny/src/lib/routes/scans.js (hunk 2 above)
Comment (swashcap) [inline packages/steelpenny/src/lib/routes/scans.js]: Disable scoped in 93dd84de27989f549f4727cc79f91eb17ef2e082.

File: packages/steelpenny/src/lib/routes/scans.js (hunk 1 above)
Comment (swashcap) [inline packages/steelpenny/src/lib/routes/scans.js]: Readability improved in 0f60f3fc911a1778e72c36f7dbf5529a2ba26eb0.

File: packages/steelpenny/src/lib/routes/scans.js [hunk 5]
```diff
@@ -60,28 +37,39 @@ exports.register = (server, options, next) => {
...
+         * @return {Promise} a promise that resolves to an array of studyIds
+         */
+        const getReadScanStudies = username => bluebird.promisify(
+          relations.study)('What can %s read_Scan from?', username);
+
+        if (!isEmpty(request.query)) {
+          return scanController.get(request, reply);
         }
+
+        getReadScanStudies(request.auth.credentials.username)
```
Comment (swashcap) [inline packages/steelpenny/src/lib/routes/scans.js]: Updated in 5477abc99457e89e6ef5f3ae322d832bcb8f0f6f.

**Comment by swashcap** [comment]
e5061b8ad164a3a81360d55280f1f4cfc4009e1b moves the `GET /scans` logic into the controller, which cleans up the route a bit. Ready for review!

File: packages/steelpenny/src/lib/controllers/scans.js [hunk 6]
```diff
@@ -1,74 +1,135 @@
...
+  tags: ['api', 'scans'],
+  notes: 'query parameters study_id and ursi are used to restrict output',
+  description: 'Returns a collection of scans',
+  plugins: {
+    'hapi-swagger': { nickname: 'get' },
+  },
+  validate: {
+    query: joi.object().keys({
+      studyId: joi.number(),
+      ursi: joi.string().min(9).max(9),
```
Comment (drewlandis) [inline packages/steelpenny/src/lib/controllers/scans.js]: is there a `.length` method for joi?

File: packages/steelpenny/src/lib/controllers/scans.js [hunk 7]
```diff
@@ -1,74 +1,135 @@
...
+    reply
+  ) {
+    const Scan = bookshelf.model('Scan');
+    const formatQuery = Scan.prototype._utils.formatQuery;
+
+    /**
+     * get all scans in the specified studies
+     * @param  {number} studies array of studyIds (int)
+     * @param  {boolean} returnJson return JSON instead of raw collection
+     * @return {Promise} A Promise that resolves to json or collection of scans
```
Comment (drewlandis) [inline packages/steelpenny/src/lib/controllers/scans.js]: `@returns`

File: packages/steelpenny/src/lib/controllers/scans.js [hunk 8]
```diff
@@ -1,74 +1,135 @@
...
+     */
+    const readAllScansInStudies = (creds, studies) => new Scan()
+      .query('whereIn', 'study_id', studies)
+      .readAll(creds);
+
+    const { promisify } = require('bluebird'); // eslint-disable-line global-require
+
+    /**
+     * Get the studies that the user is allowed to read scans from
+     * @return {Promise} a promise that resolves to an array of studyIds
```
Comment (drewlandis) [inline packages/steelpenny/src/lib/controllers/scans.js]: `@returns`

File: packages/steelpenny/src/lib/controllers/scans.js (hunk 6 above)
Comment (swashcap) [inline packages/steelpenny/src/lib/controllers/scans.js]: Good call. Looks like [there is a `joi.string().length` method](https://github.com/hapijs/joi/blob/master/API.md#stringlengthlimit-encoding).

File: packages/steelpenny/src/lib/controllers/scans.js [hunk 9]
```diff
@@ -1,49 +1,135 @@
...
+      studyId: joi.number(),
+      ursi: joi.string().min(9).max(9),
+    }).label('UserScanRequest'),
+  },
+  response: {
+    schema: joi.array().items(internals.schema).label('Scans'),
+  },
+
+  /**
+   * Get scan(s)
```
Comment (swashcap) [inline packages/steelpenny/src/lib/controllers/scans.js]: Clarified docblock in 44881accbae90bf42c709086e750f61ed18c8ef4

File: packages/steelpenny/src/lib/controllers/scans.js (hunk 6 above)
Comment (swashcap) [inline packages/steelpenny/src/lib/controllers/scans.js]: Updated in ee3836e9b5e361ed1766bd27862e037059958c09

**Comment by swashcap** [comment]
4b7d2a406b846b3c9219afae383e57cd28b5bb31 is the big refactor. Should be good to go! 🚀

**Comment by swashcap** [comment]
«bump»

**Pull request merged**